File: year2/mpp/java/MotorcycleClientServer/MotorcycleServer/src/main/java/networking/AbstractServer.java
package networking;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import services.ServerException;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

public abstract class AbstractServer {
    private int port;
    private ServerSocket server = null;
    private static final Logger LOGGER = LogManager.getLogger(AbstractServer.class.getName());

    public AbstractServer(int port){
        LOGGER.traceEntry("setting server port " + port);
        this.port = port;
    }

    public AbstractServer() {
    }

    public void setPort(int port) {
        this.port = port;
    }

    public void setServer(ServerSocket server) {
        this.server = server;
    }

    public void start() throws ServerException {
        try {
            LOGGER.info("starting server");
            server = new ServerSocket(port);
            while (true) {
                LOGGER.info("waiting for clients");
                Socket client = server.accept();
                LOGGER.info("client connected");
                processRequest(client);
            }
        } catch (IOException e) {
            LOGGER.warn("starting server error {}", e.getMessage());
            throw new ServerException("starting server error " + e);
        } finally {
            stop();
        }
    }

    protected abstract void processRequest(Socket client);

    public void stop() throws ServerException {
        try {
            LOGGER.info("closing server");
            server.close();
        } catch (IOException e) {
            LOGGER.warn("closing server error");
            throw new ServerException("closing server error " + e);
        }
        LOGGER.traceExit();
    }
}

File: year2/mpp/c#/MotorcycleClientServer/MotorcycleServer/Networking/AbstractServer.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Remoting;

namespace MotorcycleServer.Networking
{
    public abstract class AbstractServer
    {
        // protected static readonly ILog Logger = LogManager.GetLogger("AbstractServer");
        
        private TcpListener server;
        private String host;
        private int port;

        public AbstractServer(String host, int port)
        {
            // Logger.InfoFormat("creating server with host {0} and port {1}...", host, port);
            this.host = host;
            this.port = port;
        }

        public void Start()
        {
            try
            {
                // Logger.Info("starting server...");
                IPAddress address = IPAddress.Parse(host);
                IPEndPoint endPoint = new IPEndPoint(address, port);
                server = new TcpListener(endPoint);
                server.Start();
                while (true)
                {
                    // Logger.Info("waiting for clients ...");
                    Console.WriteLine("waiting for clients ...");
                    TcpClient client = server.AcceptTcpClient();
                    // Logger.Info("client connected ...");
                    Console.WriteLine("client connected ...");
                    ProcessRequest(client);
                }
            }
            catch (Exception exception)
            {
                // Logger.Warn("starting server failed");
                throw new ServerException(exception.Message);
            }
            finally
            {
                Stop();
            }
        }

        public void Stop()
        {
            try
            {
                // Logger.Info("stopping server...");
                server.Stop();
            }
            catch (Exception exception)
            {
                // Logger.Warn("stopping server failed");
                throw new ServerException(exception.Message);
            }
        }
        
        public abstract void ProcessRequest(TcpClient client);
    }
}